### PR TITLE
Tblgen fortran ones

### PR DESCRIPTION
**Pull request by ZuseZ4**
post weekend fixes for fortran BLAS usage.

### CODE REVIEW

File: enzyme/Enzyme/Utils.cpp [hunk 1]
```diff
@@ -2038,3 +2183,110 @@ llvm::FastMathFlags getFast() {
     f.set();
   return f;
 }
+
+void addValueToCache(llvm::Value *arg, bool cache_arg, llvm::Type *ty,
+                     llvm::SmallVector<llvm::Value *, 2> &cacheValues,
+                     llvm::IRBuilder<> &BuilderZ, llvm::Twine name) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:2189]: `Twine` should always be passed as a const reference

```suggestion
void addValueToCache(llvm::Value *arg, bool cache_arg, llvm::Type *ty,
                     llvm::SmallVector<llvm::Value *, 2> &cacheValues,
                     llvm::IRBuilder<> &BuilderZ, const llvm::Twine &name) {
```

File: enzyme/Enzyme/Utils.cpp [hunk 2]
```diff
@@ -2038,3 +2183,110 @@ llvm::FastMathFlags getFast() {
...
+#endif
+  cacheValues.push_back(arg);
+}
+
+void extractValueFromCache(llvm::Value *arg, bool cache_arg,
+                           llvm::Value *true_arg, llvm::Type *ty,
+                           llvm::Value *cacheval, unsigned cacheidx,
+                           DerivativeMode Mode,
+                           llvm::IRBuilder<> &allocationBuilder,
+                           llvm::IRBuilder<> &Builder2) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:2207]: nit: In .cpp files we usually use namespace llvm, so we can avoid the llvm namespace prefix here

File: enzyme/Enzyme/Utils.cpp [hunk 3]
```diff
@@ -2038,3 +2183,110 @@ llvm::FastMathFlags getFast() {
...
+  //  true_arg = Builder2.CreatePointerCast(alloc, PointerType::getUnqual(ty));
+  //  // old: Builder2.CreatePointerCast(alloc,
+  //  call.getArgOperand(0)->getType()); arg = true_arg; cacheidx++;
+}
+
+// julia_decl null means not julia decl, otherwise it is the integer type needed
+// to cast to
+llvm::Value *to_blas_callconv(IRBuilder<> &B, llvm::Value *V, bool byRef,
+                              IntegerType *julia_decl,
+                              IRBuilder<> &entryBuilder, llvm::Twine name) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:2235]: Twine should always be passed as a const reference

File: enzyme/Enzyme/Utils.cpp [hunk 4]
```diff
@@ -2038,3 +2183,110 @@ llvm::FastMathFlags getFast() {
...
+                  B.CreateICmpEQ(V, ConstantInt::get(V->getType(), 'n')),
+                  ConstantInt::get(V->getType(), 't'),
+                  ConstantInt::get(V->getType(), 0)))));
+  return out;
+}
+
+llvm::Value *transpose(llvm::IRBuilder<> &B, llvm::Value *V, bool byRef,
+                       llvm::IntegerType *julia_decl,
+                       llvm::IRBuilder<> &entryBuilder,
+					   llvm::Twine name) {
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.cpp:2268]: Twine should always be passed as a const reference

File: enzyme/Enzyme/Utils.h [hunk 5]
```diff
@@ -1609,6 +1626,29 @@ static inline bool containsOnlyAtMostTopBit(const llvm::Value *V,
   return false;
 }
 
+void addValueToCache(llvm::Value *arg, bool cache_arg, llvm::Type *ty,
+                     llvm::SmallVector<llvm::Value *, 2> &cacheValues,
+                     llvm::IRBuilder<> &BuilderZ, llvm::Twine = "");
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.h:1631]: Twine should always be passed as a const reference

File: enzyme/Enzyme/Utils.h [hunk 6]
```diff
@@ -1609,6 +1626,29 @@ static inline bool containsOnlyAtMostTopBit(const llvm::Value *V,
...
+                           llvm::Value *cacheval, unsigned cachidx,
+                           DerivativeMode Mode,
+                           llvm::IRBuilder<> &allocationBuilder,
+                           llvm::IRBuilder<> &Builder2);
+
+// julia_decl null means not julia decl, otherwise it is the integer type needed
+// to cast to
+llvm::Value *to_blas_callconv(llvm::IRBuilder<> &B, llvm::Value *V, bool byRef,
+                              llvm::IntegerType *julia_decl,
+                              llvm::IRBuilder<> &entryBuilder, llvm::Twine="");
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.h:1643]: Twine should always be passed as a const reference

File: enzyme/Enzyme/Utils.h [hunk 7]
```diff
@@ -1609,6 +1626,29 @@ static inline bool containsOnlyAtMostTopBit(const llvm::Value *V,
...
+// first one assume V is an Integer
+llvm::Value *transpose(llvm::IRBuilder<> &B, llvm::Value *V);
+// secon one assume V is an Integer or a ptr to an int (depends on byRef)
+llvm::Value *transpose(llvm::IRBuilder<> &B, llvm::Value *V, bool byRef,
+                       llvm::IntegerType *IT, llvm::IRBuilder<> &entryBuilder,
+                       llvm::Twine name);
```
Comment (tgymnich) [inline enzyme/Enzyme/Utils.h:1649]: Twine should always be passed as a const reference

**Comment by wsmoses** [comment]
@ZuseZ4 does this add new functionality or can it be closed now?

**Pull request closed without merging**